### ISSUE TITLE
[Bug] Distributed L2 cache caching tokens under wrong user

### ISSUE THREAD

**Issue by councilwisenick** [AzureAD/microsoft-authentication-library-for-dotnet#4359]
### Library version used

4.56.0

### .NET version

.NET Framework 4.6.2

### Scenario

ConfidentialClient - web site (AcquireTokenByAuthCode)

### Is this a new or an existing app?

The app is in production, I haven't upgraded MSAL, but started seeing this issue

### Issue description and reproduction steps

I recently noticed a strange issue when implementing a Redis distributed token cache which is causing user tokens to expire and not be refreshed leading to errors. What I am aiming for is one token cache per user being serialised to Redis.

If I watch the Redis server I can see keys are created for each user when they sign in (using AcquireTokenByAuthorizationCode) in the "{oid}.{tid}" format for that user. These are saved with the user's token inside them, but the data inside the cache will include all other users that have been cached by the web app as well. For example, assuming a totally empty cache:

- User A signs in, a new key is created in Redis for user A containing only that user's token
- User B signs in, a new key is created in Redis for user B, but now the new key contains both user A and user B's tokens
- As more users connect the cache seems to shift around and lookups for refresh tokens start failing because they're not in the L2 cache

Am I misunderstanding something fundamental here, or is something going wrong?

### Relevant code snippets

```csharp
//Configuration for the redis cache:

IConfidentialClientApplication clientapp = ConfidentialClientApplicationBuilder.Create(ClientId)
                  .WithClientSecret(AppKey)
                  .WithRedirectUri(ConfigurationManager.AppSettings["pw:SiteUrlAddress"])
                  .WithAuthority(new Uri("https://login.microsoftonline.com/organizations/v2.0/"))
                  .Build();

            clientapp.AddDistributedTokenCache(services =>
            {
                services.AddStackExchangeRedisCache(options =>
                {
                    options.Configuration =  GetRedisConnectionString();
                    options.InstanceName = ConfigurationManager.AppSettings["redis:tokenCachePrefix"];
                });
            });

// Initial login:
        private static async Task OnAuthorizationCodeReceived(AuthorizationCodeReceivedNotification context)
        {
            // .. snip
            
var authenticationResult = await ClaimHelper.ClientApplication.AcquireTokenByAuthorizationCode(new[] { ClaimConstants.AllScopes }, context.Code).ExecuteAsync();
        
            // .. snip
        }


// Subsequent lookup for tokens:
var account = await ClaimHelper.ClientApplication.GetAccountAsync(AccountIdentifier);
            var result = await ClientApplication.AcquireTokenSilent(new[] { ClaimConstants.ApiScope }, account).ExecuteAsync();
```

### Expected behavior

What I expected is that there should be a completely separate token cache for each user connecting, and so each key in Redis should include a single token for that corresponding user.

### Identity provider

Microsoft Entra ID (Work and School accounts and Personal Microsoft accounts)

**Comment by bgavrilMS** [comment]
@pmaytak - please take a look at this, seems serious.

**Comment by councilwisenick** [comment]
Hi, 

Today I seem to have solved this issue. 

It seems to have been caused by some combination of packages I was using. In particular it seems like it was using the deprecated redis package over the StackExchange.Redis and Microsoft.Extensions.Caching.StackExchangeRedis packages and there was a mismatch of versions across a couple of different projects involved. After I changed over to the supported packages and consolidated all the versions to the latest, the issue appears to be resolved. 

I have pushed a fix and will continue to monitor today to make sure the issue has gone away.
